src/hardware/shared_adc.rs

```diff
@@ -12,24 +12,64 @@
 /// [AdcChannel]. The [AdcChannel]'s ownership can then be moved to any required drivers.
 ///
 /// ## Synchronization
-/// Currently, the sharing of the ADC peripheral across multiple priority levels is managed by the
-/// use of a critical section.
+/// If the multiple priorities utilize the ADC that results in resource pre-emption, pre-emption is
+/// protected against through the use of an atomic bool. Attempting to utilize the ADC from a
+/// higher priority level while it is in use at a lower level will result in a [AdcError::InUse].
 use embedded_hal::adc::{Channel, OneShot};
 use stm32h7xx_hal as hal;
 
+struct AdcMutex<Adc> {
+    in_use: core::sync::atomic::AtomicBool,
+    adc: core::cell::RefCell<hal::adc::Adc<Adc, hal::adc::Enabled>>,
+}
+
+impl<Adc> AdcMutex<Adc> {
+    pub fn new(adc: hal::adc::Adc<Adc, hal::adc::Enabled>) -> Self {
+        Self {
+            adc: core::cell::RefCell::new(adc),
+            in_use: core::sync::atomic::AtomicBool::new(false),
```

Comment:
Isn't there a generic mutex no_std crate that does this? If not, shouldn't there be?
Shouldn't this be an `impl` of https://rahix.github.io/embedded-hal/embedded_hal/mutex/trait.RwMutex.html ?

Reply:
Maybe there's other structurs in Spinning that do this. Give me a moment.